### PR TITLE
New logger connector code

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Introduces the new code for #7631

Next PRs:
- Update the loops to use this
- Delete previous code

cc: @SkafteNicki please review our heavy use of `Metric` in `result_new.py`

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+                self.trainer.logger.agg_and_log_metrics(scalar_metrics, step=step)
+                self.trainer.logger.save()
+
+            self._logged_metrics.update(scalar_metrics)
+
+    """
+    Evaluation metric updates
+    """
+
+    def prepare_eval_loop_results(self, metrics: Mapping[str, _METRIC]) -> None:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:120]: Could be move to utils.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+                })
+                print('-' * 80)
+
+        results = self.eval_loop_results
+        # clear mem
+        self.eval_loop_results = []
+        return results
+
+    @property
+    def evaluation_log_step(self) -> Optional[int]:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:169]: Same here.

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:169]: Let's keep `update_{state}_{step|epoch}_metrics` and hooks.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        # add the metrics to the loggers
+        assert self._epoch_end_reached
+        metrics = self.metrics[MetricSource.LOG]
+        if metrics:
+            self.log_metrics(metrics)
+
+        # reset result collection for next epoch
+        self.trainer.result_collection.reset(metrics=True)
+
+    def teardown(self):
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:234]: is that something we should do in the loops maybe? 
just saying, no change requersted

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        self._split_idx: Optional[int] = None
+
+    def on_trainer_init(
+        self, logger: LightningLoggerBase, flush_logs_every_n_steps: int, log_every_n_steps: int,
+        move_metrics_to_cpu: bool
+    ) -> None:
+        self.configure_logger(logger)
+        self.trainer.flush_logs_every_n_steps = flush_logs_every_n_steps
+        self.trainer.log_every_n_steps = log_every_n_steps
+        self.trainer.move_metrics_to_cpu = move_metrics_to_cpu
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:54]: would be great to have logger connector own these properties and make them read only forward properties on the trainer. 
could be done in a follow up, what do you say?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+from pytorch_lightning.loggers import LightningLoggerBase, LoggerCollection, TensorBoardLogger
+from pytorch_lightning.trainer.connectors.logger_connector.result_new import _METRIC, MetricSource
+from pytorch_lightning.trainer.states import RunningStage, TrainerFn
+from pytorch_lightning.utilities import DeviceType
+from pytorch_lightning.utilities.metrics import metrics_to_scalars
+from pytorch_lightning.utilities.types import _EVALUATE_OUTPUT
+
+
+# TODO(@carmocca): Remove `New` suffix
+class LoggerConnectorNew:
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:31]: is this replacement for the old one?

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:31]: pr description says yes, old will be removed :)

**Review by tchaton — APPROVED** [review]
LGTM ! Great job @carmocca

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:31]: yes

File: requirements.txt [hunk 6]
```diff
@@ -7,7 +7,7 @@ tqdm>=4.41.0
 PyYAML>=5.1,<=5.4.1
 fsspec[http]>=2021.05.0, !=2021.06.0
 tensorboard>=2.2.0, !=2.5.0  # 2.5.0 GPU CI error: 'Couldn't build proto file into descriptor pool!'
-torchmetrics>=0.2.0
+torchmetrics>=0.3.2
```
[the remark this answers is not in the document]
Comment (carmocca) [reply requirements.txt:10]: we heavily rely on `Metric` for the new result class and the newer versions have important changes we need

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:31]: and why not remove it in this PR? does it require some extra work?

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:31]: first the new connector needs to be integrated into the rest of the code base, ie., loops. the same will be true for our loop refactors. I think the approach that Carlos takes is the best to brake down the complexity :)

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:120]: Why? It's tied to evaluation

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:169]: Done

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:234]: The traine will call this:

```python
    def _post_dispatch(self):
        self.accelerator.post_dispatch(self)
        self.accelerator.teardown()
        self.logger_connector.teardown()
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:57]: Can be easily done for `flush_logs_every_n_steps` as it is not used anywhere else.

The other two are used in other places. What's the policy then?

Another concern is that users might be expecting/using them. It's is natural to expect the attribute when it was passed to the `Trainer` constructor

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:31]: Exactly, if we remove the old one here, we also need to update all the loops and a ton of tests in this same PR

(which would equal to what we have in #7631)

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:57]: > The other two are used in other places. What's the policy then?

I'm saying to make them read only accessible on the trainer. Either way, no matter where they live I believe the state of a class should be managed ONLY by that class or by message passing, i.e., a method call.

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:57]: Example: https://github.com/PyTorchLightning/pytorch-lightning/pull/7437/files

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    def progress_bar_metrics(self) -> Dict[str, float]:
+        if self.trainer.results:
+            metrics = self.metrics[MetricSource.PBAR]
+            self._progress_bar_metrics.update(metrics)
+        return self._progress_bar_metrics
+
+    def teardown(self):
+        self.trainer.train_loop.results.cpu()
+        self.trainer.evaluation_loop._val_results.cpu()
+        self.trainer.evaluation_loop._test_results.cpu()
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:310]: @awaelchli do you think each loop should run this on their teardown?
instead of the logger connector doing it

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:311]: these results are the resultscollection right? I think yes, if the loops own this results collection I think they should handle the teardown

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+        # turn all tensors to scalars
+        scalar_metrics = metrics_to_scalars(metrics)
+
+        if "step" in scalar_metrics and step is None:
+            step = scalar_metrics.pop("step")
+
+        elif step is None:
+            # added metrics by Lightning for convenience
+            scalar_metrics['epoch'] = self.trainer.current_epoch
+            step = self.trainer.global_step
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:109]: There's a lot of Codecov warnings.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:311]: Okay. Do you think I should do that change here or you do it when it's all merged?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:109]: This is expected considering we haven't integrated the loops to use these new files.

You can take a peek at #7631 for the full integration

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:109]: This PR includes a few unit tests for the code in `Result` but `LoggerConnectorNew` is completely unused

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    def should_flush_logs(self) -> bool:
+        should_flush = (self.trainer.global_step + 1) % self.trainer.flush_logs_every_n_steps == 0
+        return should_flush or self.trainer.should_stop
+
+    @property
+    def should_update_logs(self) -> bool:
+        should_log_every_n_steps = (self.trainer.global_step + 1) % self.trainer.log_every_n_steps == 0
+        return should_log_every_n_steps or self.trainer.should_stop
+
+    def configure_logger(self, logger: LightningLoggerBase) -> None:
```
Comment (ethanwharris) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:69]: Should this be typed as `Union[bool, Iterable, LightningLoggerBase]`?

**Review by ethanwharris — APPROVED** [review]
Nice! LGTM :smiley: minor comment

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+                self.log_metrics(log_metrics)
+
+        self._prepare_eval_loop_results(metrics[MetricSource.CALLBACK])
+
+        # log results of evaluation
+        if (
+            self.trainer.state.fn != TrainerFn.FITTING and self.trainer.evaluating and self.trainer.is_global_zero
+            and self.trainer.verbose_evaluate
+        ):
+            print('-' * 80)
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:194]: Required?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py]: Added to next PR :)

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 8 above)
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:109]: n00b questions: 
- why does step indicate we should epoch to scalar metrics? 
- what happens if epoch was already in scalar metrics as another value? if the user does `self.log("epoch", ...)` ?

File: pytorch_lightning/trainer/connectors/logger_connector/result_new.py [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+
+# re-define the ones from pytorch_lightning.utilities.types without the `Number` type
+_METRIC = Union[Metric, torch.Tensor]
+_METRIC_COLLECTION = Union[_METRIC, Mapping[str, _METRIC]]
+
+
+class MetricSource(LightningEnum):
+    CALLBACK = "callback"
+    PBAR = "pbar"
+    LOG = "log"
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/result_new.py:38]: are these metric sources, or are they destinations? 

in the future, i can imagine we want to configure multiple loggers, but want to log metrics to only some of them. how could this be extended for that use case?

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result_new.py:38]: > are these metric sources, or are they destinations?

Mmm depends on how you think about it.

Do you `self.log(prog_bar=True)` into the progress bar metrics (destination) or are the progress bar metrics gathered from the `prog_bar=True` logged values (source)?

If you have an idea for a less confusing name, we can rename without issues 👍 

> in the future, i can imagine we want to configure multiple loggers, but want to log metrics to only some of them. how could this be extended for that use case?

In that case, I think it's easier for us to filter at the logger level. `_Metadata` could have a field for the target logger

File: pytorch_lightning/trainer/connectors/logger_connector/result_new.py [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+        :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step`
+        """
+        return self.get('_extra', {})
+
+    @extra.setter
+    def extra(self, extra: Mapping[str, Any]) -> None:
+
+        def check_fn(v):
+            if v.grad_fn is not None:
+                raise MisconfigurationException(f'You returned a tensor with `grad_fn`. The extra values are {extra}')
```
Comment (kandluis) [inline pytorch_lightning/trainer/connectors/logger_connector/result_new.py:265]: Could you explain what this checks for? Is it a misconfiguration for users to return tensors with gradients enabled?

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result_new.py:265]: IIRC the previous implementation detached extras (silently) internally, however, I feel like it's better for the users to know they are passing tensors with graphs. This facilitates catching potential errors on the user side and it's not much to ask the user to detach if necessary.

Do you think this constraint should be relaxed?

Comment (kandluis) [reply pytorch_lightning/trainer/connectors/logger_connector/result_new.py:265]: Not entirely sure - just noticed that quite a few of our users were relying on this (silent) behavior and will see exceptions with the latest version of the code here. 

Could you share an example of how/what a users should do now? Is the idea they want to call `detach` on non-loss tensors?

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result_new.py:265]: > Is the idea they want to call detach on non-loss tensors?

Exactly:

```python
def training_step(...):
    ...
    # `something_else` is considered an "extra"
    return {'loss': loss, 'something_else': something_else.detach()}
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+            self.trainer.logger = TensorBoardLogger(
+                save_dir=self.trainer.default_root_dir, version=version, name='lightning_logs'
+            )
+        elif logger is False:
+            self.trainer.logger = None
+        else:
+            if isinstance(logger, Iterable):
+                self.trainer.logger = LoggerCollection(logger)
+            else:
+                self.trainer.logger = logger
```
Comment (yifuwang) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:83]: nit: maybe consider explicitly checking the type of `logger` so type mismatches get caught earlier?

```
elif isinstance(logger, LightningLoggerBase):
    self.trainer.logger = logger
else:
    raise TypeError(...)
```
